### DISCUSSION TITLE
Building a simplified controller

### DISCUSSION THREAD

**Discussion by ghost** [actions/actions-runner-controller#2069] [Ideas]
So I don't know how this could be contributed back to ARC, but if this can be brought back into ARC after I prototype it; that would be great. Regardless of if we can home this in ARC or if it deviates too much, I'm open to discussing it while I prototype it.

# Overview
A simplified controller could just spawn a Kubernetes Job resource per individual GH Actions Job.

This would eliminating the need for the level of access required to use ARC today and simplifies the configuration. This would work in a single namespace for a user that doesn't have cluster admin access.

# Implementing
A runner containers that will wait for a job, do a single job, and self-terminate would be necessary. I've began implementing this here since I couldn't find one from GH themselves: https://github.com/protosam/gh-actions-runner

I think the runner has logic for handling jobs across entire accounts or for specific repos. Through some under-documented APIs mentioned [here](https://github.com/actions/runner/blob/main/docs/checks/actions.md). Trying to leverage whatever mechanisms the runner uses is part of the end goal.

At the very least, polling specific repos for queued jobs `https://api.github.com/repos/<owner>/<repo>/actions/runs?status=queued` and spawning jobs for self-hosted runs seems like a viable solution. 

# Tunables
Considering that the model is to spawn one runner per job queued, `actions-runner-spawner` should be tunable to throttle how many jobs can run at any given time.

The configuration will come from a ConfigMap for single namespace deployments.

Config will include knobsto tweak settings per repository like:
- runner container image
- runner environment variables
- namespace of the job
- pod annotations
- pod labels
- pod service account
- container resource requests
- container security context

**Comment by mumoshu** [comment]
For anyone interested, please see my comment at https://github.com/actions-runner-controller/actions-runner-controller/issues/2061#issuecomment-1336573400 too. TL;DR; with the current model of `workflow_job` webhook events, naively implementing it results in missing webhook events ending up with with the inconsistency between the new impl and GitHub Actions forever. You need a `resync` mechanism. ARC as of today uses `minReplicas` so that it can eventually become in sync with Actions as it receives workflow_job events.

**Reply by ghost** [reply]
Does this mean that a loop as described below will not work?

- `queued_runs` = slice of queued runs from  `api.github.com/repos/<owner>/<repo>/actions/runs?status=queued`.
- `pending_runs` = slice of waiting runs from  `api.github.com/repos/<owner>/<repo>/actions/runs?status=in_progress`.
- `runner_jobs_minimum_required` = `len(queued_runs) + len(pending_count)`
- `runner_job_count` = number of batch/v1 Job runners in `Running` status
- `need_runners` = `runner_jobs_minimum_required - runner_job_count`
- if `runner_job_count > 0` then create `runner_job_count` batch/v1 Jobs with `ghcr.io/protosam/gh-actions-runner` to process them

That would be my un-ideal solution, my preference is to have the controller processing what appears to be feeds. 

From the logs I noticed these paths and am planning to record what they are doing. The logs lead me to believe watchable streams are provided in this api.
- https://pipelines.action.githubusercontent.com/<some-key?>
- https://pipelines.action.githubusercontent.com/<some-key?>/_apis/dstributedtask/pools/<1 | whatever this number represents>/messages?sessionId=<session uuid here>

**Comment by mumoshu** [comment]
> Does this mean that a loop as described below will not work?

Any pull-based model is very likely to suffer from GitHub API rate limit when you are going to manage runners for many repositories, organizations, and enterprises. At least that's what happened when we implemented pull-based autoscaling a year or two ago. We then implemented HTTP caching. It reads their Cache-Control response header which instructs you to cache the responses for 60 seconds, which means the scaling operation can be delayed by at most 60 seconds. If that's fine for your use-case, pull-based would work.
In terms of ARC, it needs to support variety of use-cases so the pull-based method can't be the only supported option.

> That would be my un-ideal solution, my preference is to have the controller processing what appears to be feeds.

Me too. And my understanding is that GitHub Actions needs to provide another API that looks like a "durable and replayable log of workflow_job events" I have mentioned in https://github.com/actions-runner-controller/actions-runner-controller/issues/2061#issuecomment-1336573400. The feeds needs to be durable and replayable. If it's non-durable or non-replayable, we need the ability to work-around missing webhook events.
